xc/device.ml

```diff
@@ -607,12 +607,36 @@ let qemu_media_change ~xs device _type params =
 
 	if is_upstream_qemu device.frontend.domid
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
+		match params with
+		| "" ->
+			let qmp_cmd = Command(None, Eject (cd, Some true)) in
+			qmp_write device.frontend.domid qmp_cmd
+		| _ ->
+			try
+				let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-libxl-%d" device.frontend.domid) in
+				let fd_of_c = Qmp_protocol.to_fd c in
+				let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
+				finally
+					(fun () -> ignore(Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_cd))
+					(fun () -> Unix.close fd_of_cd);
+
+				let qmp_cmd = Command (None, Add_fd None) in
+				Qmp_protocol.negotiate c;
+				Qmp_protocol.write c qmp_cmd;
+				let qmp_change_cd_cmd, qmp_remove_fdset_cmd = match Qmp_protocol.read c with
+				| Success (None, Fd_info fd_info) -> Command (None, Blockdev_change_medium (cd, "/dev/fdset/" ^ (string_of_int fd_info.fdset_id))), Command (None, Remove_fd fd_info.fdset_id)
+				| _ -> raise (Internal_error (Printf.sprintf "Get unexpected result after sending Qmp message: %s" (string_of_message qmp_cmd)))
+				in
+				Qmp_protocol.write c qmp_change_cd_cmd;
+				Qmp_protocol.write c qmp_remove_fdset_cmd;
+				Qmp_protocol.close c
```

Comment:
With things like this it is often worthwhile to have something like
```
Qmp_protocol.with_connection
  (Printf.sprintf ...)
  (fun c ->
    let fd_of_cd = ...
    ...
  )
```
where that function is defined as something like
```
let with_connection path f =
  let c = Qmp_protocol.connect path in
  finally
    (fun () -> f c)
    (fun () -> Qmp_protocol.close c)
```